File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/components/CoachMarkOverlay.kt
package com.rork.guidestreamtvandroid.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.CompositingStrategy
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay
import com.rork.guidestreamtvandroid.data.repository.CoachMark
import com.rork.guidestreamtvandroid.data.repository.CoachMarkManager
import com.rork.guidestreamtvandroid.ui.theme.BrandOrange
import com.rork.guidestreamtvandroid.ui.theme.Navy
import kotlin.math.roundToInt

private val ScrimColor = Navy.copy(alpha = 0.57f)
private val RingColor = BrandOrange
private val CardColor = BrandOrange
private val CardTextColor = Navy
private val CardBodyColor = Navy.copy(alpha = 0.72f)
private val InactiveDotColor = Navy.copy(alpha = 0.30f)

/**
 * Spotlight overlay that dims the screen, cuts holes around one or two
 * real UI elements, and shows a callout card. Mirrors iOS CoachMarkOverlay.
 *
 * Drawn with a Canvas inside Modifier.graphicsLayer with
 * CompositingStrategy.Offscreen so BlendMode.Clear punches holes correctly.
 */
@Composable
fun CoachMarkOverlay(
    manager: CoachMarkManager,
    modifier: Modifier = Modifier,
    topInset: Dp = 12.dp,
    bottomInset: Dp = 12.dp,
) {
    // Watchdog: if a scroll request never settles (no host handles it),
    // force-settle after 1.2s so the overlay can never hang invisibly.
    LaunchedEffect(manager.isShowing, manager.currentMark?.key) {
        if (!manager.isShowing) return@LaunchedEffect
        delay(1200)
        if (manager.isShowing && !manager.scrollSettled) {
            manager.markScrollSettled()
        }
    }

    Box(modifier = modifier.fillMaxSize()) {
        CoachMarkSpotlight(
            manager = manager,
            topInset = topInset,
            bottomInset = bottomInset,
        )

        AnimatedVisibility(
            visible = manager.completionToastVisible,
            enter = slideInVertically { it } + fadeIn(),
            exit = slideOutVertically { it } + fadeOut(),
            modifier = Modifier.align(Alignment.BottomCenter),
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(10.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .padding(horizontal = 20.dp, vertical = 0.dp)
                    .padding(bottom = 86.dp)
                    .background(BrandOrange, RoundedCornerShape(14.dp))
                    .padding(horizontal = 16.dp, vertical = 12.dp),
            ) {
                Icon(
                    imageVector = Icons.Filled.CheckCircle,
                    contentDescription = null,
                    tint = Navy,
                    modifier = Modifier.size(18.dp),
                )
                Text(
                    text = "You're set. Tap any poster and we'll show you the rest.",
                    fontSize = 13.sp,
                    lineHeight = 19.sp,
                    fontWeight = FontWeight.Medium,
                    color = Navy,
                )
            }
        }
    }
}

/**
 * The actual spotlight surface: dimmed scrim with holes, pulse rings and
 * callout card. Kept identical to the previous implementation.
 */
@Composable
private fun CoachMarkSpotlight(
    manager: CoachMarkManager,
    topInset: Dp,
    bottomInset: Dp,
) {
    if (!manager.isShowing) return
    val mark = manager.currentMark ?: return
    if (!manager.scrollSettled) return

    val validRects = mark.targetKeys.mapNotNull { key ->
        manager.measuredRects[key]?.takeIf { !it.isEmpty }?.let { key to it }
    }

    if (validRects.isEmpty()) return

    val cutoutRects = validRects.map { it.second }
    val firstRect = validRects.first().second
    val density = LocalDensity.current
    val configuration = LocalConfiguration.current
    val screenWidthPx = with(density) { configuration.screenWidthDp.dp.toPx() }
    val screenHeightPx = with(density) { configuration.screenHeightDp.dp.toPx() }
    val screenRect = Rect(offset = Offset.Zero, size = Size(screenWidthPx, screenHeightPx))

    val padding8 = with(density) { 8.dp.toPx() }
    val padding12 = with(density) { 12.dp.toPx() }
    val padding10 = with(density) { 10.dp.toPx() }
    val topInsetPx = with(density) { topInset.toPx() }
    val bottomInsetPx = with(density) { bottomInset.toPx() }
    val radius14 = with(density) { 14.dp.toPx() }
    val stroke2 = with(density) { 2.dp.toPx() }
    val cardWidthPx = with(density) { 230.dp.toPx() }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .clickable(
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
            ) { manager.advance() }
            .graphicsLayer { compositingStrategy = CompositingStrategy.Offscreen },
    ) {
        // Scrim with holes — single Canvas draws all holes from one mask using
        // BlendMode.Clear so the underlying UI remains visible, framed only by
        // the orange pulse rings below.
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer { compositingStrategy = CompositingStrategy.Offscreen },
        ) {
            drawRect(color = ScrimColor, size = size)
            cutoutRects.forEach { rect ->
                val expanded = Rect(
                    offset = Offset(
                        (rect.left - padding8).coerceAtLeast(0f),
                        (rect.top - padding8).coerceAtLeast(0f),
                    ),
                    size = Size(
                        rect.width + padding8 * 2,
                        rect.height + padding8 * 2,
                    ),
                )
                if (mark.isCircular) {
                    val dim = maxOf(expanded.width, expanded.height)
                    val cx = expanded.center.x
                    val cy = expanded.center.y
                    drawCircle(
                        color = Color.Transparent,
                        radius = dim / 2f,
                        center = Offset(cx, cy),
                        blendMode = BlendMode.Clear,
                    )
                } else {
                    drawRoundRect(
                        color = Color.Transparent,
                        topLeft = expanded.topLeft,
                        size = expanded.size,
                        cornerRadius = CornerRadius(radius14, radius14),
                        blendMode = BlendMode.Clear,
                    )
                }
            }
        }

        // Pulse rings around each cutout
        cutoutRects.forEach { rect ->
            PulseRing(
                rect = rect,
                isCircular = mark.isCircular,
                padding8 = padding8,
                radius14 = radius14,
                stroke2 = stroke2,
                density = density,
            )
        }

        // Callout card
        CalloutCard(
            mark = mark,
            anchorRect = firstRect,
            allRects = cutoutRects,
            screenRect = screenRect,
            index = manager.currentIndex,
            total = manager.activeTour.size,
            manager = manager,
            padding8 = padding8,
            padding10 = padding10,
            padding12 = padding12,
            topInset = topInset,
            bottomInset = bottomInset,
            cardWidthPx = cardWidthPx,
            density = density,
        )
    }
}

@Composable
private fun PulseRing(
    rect: Rect,
    isCircular: Boolean,
    padding8: Float,
    radius14: Float,
    stroke2: Float,
    density: androidx.compose.ui.unit.Density,
) {
    val transition = rememberInfiniteTransition(label = "coach_pulse")
    val pulseAlpha by transition.animateFloat(
        initialValue = 0.9f,
        targetValue = 0.15f,
        animationSpec = infiniteRepeatable(
            animation = tween(850),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "pulse_alpha",
    )

    val expanded = Rect(
        offset = Offset(
            (rect.left - padding8).coerceAtLeast(0f),
            (rect.top - padding8).coerceAtLeast(0f),
        ),
        size = Size(
            rect.width + padding8 * 2,
            rect.height + padding8 * 2,
        ),
    )

    Canvas(
        modifier = Modifier
            .offset {
                IntOffset(
                    expanded.left.roundToInt(),
                    expanded.top.roundToInt(),
                )
            }
            .size(
                width = with(density) { expanded.width.toDp() },
                height = with(density) { expanded.height.toDp() },
            ),
    ) {
        if (isCircular) {
            val dim = maxOf(size.width, size.height)
            drawCircle(
                color = RingColor.copy(alpha = pulseAlpha),
                radius = dim / 2f,
                style = Stroke(width = stroke2),
            )
        } else {
            drawRoundRect(
                color = RingColor.copy(alpha = pulseAlpha),
                topLeft = Offset.Zero,
                size = size,
                cornerRadius = CornerRadius(radius14, radius14),
                style = Stroke(width = stroke2),
            )
        }
    }
}

@Composable
private fun CalloutCard(
    mark: CoachMark,
    anchorRect: Rect,
    allRects: List<Rect>,
    screenRect: Rect,
    index: Int,
    total: Int,
    manager: CoachMarkManager,
    padding8: Float,
    padding10: Float,
    padding12: Float,
    topInset: Dp,
    bottomInset: Dp,
    cardWidthPx: Float,
    density: androidx.compose.ui.unit.Density,
) {
    val lowestBottom = allRects.maxOf { it.bottom }
    val highestTop = allRects.minOf { it.top }
    val midScreen = screenRect.center.y
    val belowCard = lowestBottom < midScreen

    var cardHeightPx by remember { mutableFloatStateOf(140f) }

    // Narrow screens must not push the card off either edge, so the nominal
    // width collapses to whatever the screen can actually hold.
    val maxCardWidthPx = (screenRect.width - padding10 * 2f).coerceAtLeast(0f)
    val actualCardWidthPx = cardWidthPx.coerceAtMost(maxCardWidthPx)
    val cardWidthDp = with(density) { actualCardWidthPx.toDp() }

    var rawX = anchorRect.center.x - actualCardWidthPx / 2f
    rawX = rawX.coerceIn(
        padding10,
        (screenRect.width - actualCardWidthPx - padding10).coerceAtLeast(padding10),
    )

    val topInsetPx = with(density) { topInset.toPx() }
    val bottomInsetPx = with(density) { bottomInset.toPx() }

    val preferredTop = if (belowCard) {
        lowestBottom + padding12
    } else {
        highestTop - padding12 - cardHeightPx
    }
    // The card is positioned manually, so nothing else keeps it on screen:
    // clamp its top edge to the safe band. Without this a tall card anchored
    // near an edge (or a target close to the bottom) runs off the display.
    val minTop = topInsetPx.coerceAtLeast(padding12)
    val maxTop = (screenRect.height - cardHeightPx - bottomInsetPx)
        .coerceAtLeast(minTop)
    val cardTop = preferredTop.coerceIn(minTop, maxTop)

    Box(
        modifier = Modifier
            .offset {
                IntOffset(
                    rawX.roundToInt(),
                    cardTop.roundToInt(),
                )
            }
            .width(cardWidthDp)
            .graphicsLayer { compositingStrategy = CompositingStrategy.Offscreen },
    ) {
        Column(
            modifier = Modifier
                .width(cardWidthDp)
                .background(CardColor, RoundedCornerShape(12.dp))
                .padding(horizontal = 14.dp, vertical = 12.dp)
                .onGloballyPositioned { coords ->
                    cardHeightPx = coords.size.height.toFloat()
                },
        ) {
            Text(
                text = mark.title,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = CardTextColor,
            )
            Spacer(Modifier.size(4.dp))
            Text(
                text = mark.body,
                fontSize = 12.sp,
                lineHeight = 18.sp,
                color = CardBodyColor,
            )
            Spacer(Modifier.size(10.dp))
            // fillMaxWidth, never fillMaxSize: inside a wrap-height Column a
            // fillMaxSize footer stretches the card to the full screen height,
            // which is what made the callout a giant orange slab.
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    for (i in 0 until total) {
                        Box(
                            modifier = Modifier
                                .size(5.dp)
                                .background(
                                    if (i == index) CardTextColor else InactiveDotColor,
                                    CircleShape,
                                ),
                        )
                    }
                }
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = "Skip",
                        fontSize = 11.sp,
                        color = CardTextColor.copy(alpha = 0.72f),
                        modifier = Modifier.clickable(
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                        ) { manager.skipTour() },
                    )
                    Spacer(Modifier.size(8.dp))
                    Text(
                        text = if (manager.currentIndex == manager.activeTour.size - 1) "Done" else "Next",
                        fontSize = 12.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = CardTextColor,
                        modifier = Modifier.clickable(
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                        ) { manager.advance() },
                    )
                }
            }
        }
    }
}

File: ios/GuideStreamTV/Views/CoachMarkOverlay.swift
//
//  CoachMarkOverlay.swift
//  GuideStreamTV
//
//  Spotlight overlay that dims the screen, cuts holes around one or two
//  real UI elements, and shows a callout card explaining them. Uses a
//  PreferenceKey carrying [String: Anchor<CGRect>] for target measurement.
//

import SwiftUI

// MARK: - PreferenceKey for anchor-based target measurement

struct CoachMarkAnchorKey: PreferenceKey {
    static var defaultValue: [String: Anchor<CGRect>] = [:]
    static func reduce(value: inout [String: Anchor<CGRect>], nextValue: () -> [String: Anchor<CGRect>]) {
        value.merge(nextValue(), uniquingKeysWith: { _, b in b })
    }
}

// MARK: - Overlay view

struct CoachMarkOverlay: View {
    let manager: CoachMarkManager
    let topInset: CGFloat
    let bottomInset: CGFloat

    init(manager: CoachMarkManager, topInset: CGFloat = 104, bottomInset: CGFloat = 104) {
        self.manager = manager
        self.topInset = topInset
        self.bottomInset = bottomInset
    }

    var body: some View {
        ZStack {
            GeometryReader { geo in
                let screen = geo.frame(in: .global)
                if manager.isShowing, let mark = manager.currentMark {
                    overlayContent(mark: mark, screen: screen)
                        .transition(.opacity)
                } else {
                    Color.clear
                }
            }
            .ignoresSafeArea()

            if manager.completionToastVisible {
                completionToast
            }
        }
        .allowsHitTesting(manager.isShowing)
        .animation(.easeInOut(duration: 0.25), value: manager.isShowing)
        .animation(.spring(response: 0.35, dampingFraction: 0.85), value: manager.completionToastVisible)
    }

    // MARK: - Completion toast

    private var completionToast: some View {
        HStack(spacing: 10) {
            Image(systemName: "checkmark.circle.fill")
                .scaledFont(size: 15, weight: .semibold)
            Text("You're set. Tap any poster and we'll show you the rest.")
                .scaledFont(size: 13, weight: .medium)
                .fixedSize(horizontal: false, vertical: true)
        }
        .foregroundStyle(Color.navy)
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(RoundedRectangle(cornerRadius: 14).fill(Color.orange))
        .shadow(color: .black.opacity(0.5), radius: 18, y: 8)
        .padding(.horizontal, 20)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
        .padding(.bottom, 86)
        .transition(.move(edge: .bottom).combined(with: .opacity))
    }

    // MARK: - Scrim + cutouts + card

    @ViewBuilder
    private func overlayContent(mark: CoachMark, screen: CGRect) -> some View {
        let rects = manager.measuredRects
        let validRects = mark.targetKeys.compactMap { key -> (String, CGRect)? in
            guard let r = rects[key], !r.isEmpty else { return nil }
            return (key, r)
        }

        if validRects.isEmpty || !manager.scrollSettled {
            // Waiting for scroll to settle or no valid frames — show nothing
            Color.clear
                .onAppear {
                    attemptSkipIfAllUnmeasurable(mark: mark)
                }
                .onChange(of: manager.measureAttempt) { _, _ in
                    attemptSkipIfAllUnmeasurable(mark: mark)
                }
        } else {
            let cutoutRects = validRects.map { $0.1 }
            ZStack {
                scrimWithHoles(rects: cutoutRects, mark: mark, screen: screen)

                ForEach(Array(cutoutRects.indices), id: \.self) { idx in
                    pulseRing(rect: cutoutRects[idx], isCircular: mark.isCircular, screen: screen)
                }

                if let firstValid = validRects.first {
                    calloutCard(
                        mark: mark,
                        anchorRect: firstValid.1,
                        allRects: cutoutRects,
                        screen: screen,
                        index: manager.currentIndex,
                        total: manager.activeTour.count
                    )
                }
            }
            .contentShape(Rectangle())
            .onTapGesture {
                manager.advance()
            }
        }
    }

    private func attemptSkipIfAllUnmeasurable(mark: CoachMark) {
        guard manager.scrollSettled, manager.measureAttempt > 0 else { return }
        let allUnmeasured = mark.targetKeys.allSatisfy {
            manager.measuredRects[$0]?.isEmpty ?? true
        }
        guard allUnmeasured else { return }
        let markKey = mark.key
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            guard manager.currentMark?.key == markKey else { return }
            guard let current = manager.currentMark else { return }
            let stillAllUnmeasured = current.targetKeys.allSatisfy {
                manager.measuredRects[$0]?.isEmpty ?? true
            }
            guard stillAllUnmeasured else { return }
            manager.skipUnmeasurableMark()
        }
    }

    // MARK: - Scrim with holes (single mask, no double-opacity)

    private func scrimWithHoles(rects: [CGRect], mark: CoachMark, screen: CGRect) -> some View {
        Canvas { context, size in
            // Fill entire screen with scrim
            context.fill(Path(CGRect(origin: .zero, size: size)),
                         with: .color(Color.navy.opacity(0.57)))
            // Cut all holes from a single mask
            context.blendMode = .destinationOut
            for rect in rects {
                let expanded = Self.clampedCutout(for: rect, in: CGRect(origin: .zero, size: size))
                if mark.isCircular {
                    let dim = max(expanded.width, expanded.height)
                    let cx = expanded.midX
                    let cy = expanded.midY
                    let r = dim / 2
                    context.fill(Path(ellipseIn: CGRect(x: cx - r, y: cy - r, width: r * 2, height: r * 2)),
                                 with: .color(.white))
                } else {
                    context.fill(RoundedRectangle(cornerRadius: 14).path(in: expanded),
                                 with: .color(.white))
                }
            }
            context.blendMode = .normal
        }
        .frame(width: screen.width, height: screen.height)
        .ignoresSafeArea()
    }

    // MARK: - Pulse ring (opacity 0.9 → 0.15 → 0.9 over 1.7s)

    @ViewBuilder
    private func pulseRing(rect: CGRect, isCircular: Bool, screen: CGRect) -> some View {
        let expanded = Self.clampedCutout(for: rect, in: CGRect(origin: .zero, size: screen.size))
        PulseRingView(expanded: expanded, isCircular: isCircular)
            .ignoresSafeArea()
    }

    /// Expands the target rect by 8pt, then clamps it so the cutout and
    /// pulse ring never bleed past the screen edges (6pt minimum margin).
    static func clampedCutout(for rect: CGRect, in bounds: CGRect) -> CGRect {
        let expanded = rect.insetBy(dx: -8, dy: -8)
        let limit = bounds.insetBy(dx: 6, dy: 6)
        let clamped = expanded.intersection(limit)
        return clamped.isNull || clamped.isEmpty ? expanded : clamped
    }

    // MARK: - Callout card

    @ViewBuilder
    private func calloutCard(
        mark: CoachMark,
        anchorRect: CGRect,
        allRects: [CGRect],
        screen: CGRect,
        index: Int,
        total: Int
    ) -> some View {
        let lowestBottom = allRects.map { $0.maxY }.max() ?? anchorRect.maxY
        let highestTop = allRects.map { $0.minY }.min() ?? anchorRect.minY
        let midScreen = screen.midY
        let cardWidth: CGFloat = 230

        // Position: 12pt below lowest cutout if above midpoint,
        // otherwise 12pt above highest cutout
        let belowCard: Bool = lowestBottom < midScreen

        let rawY: CGFloat = belowCard
            ? lowestBottom + 12 + (positionedCardSize.height / 2)
            : highestTop - 12 - (positionedCardSize.height / 2)
        let cardHeight: CGFloat = positionedCardSize.height
        let minY: CGFloat = topInset + cardHeight / 2
        let maxY: CGFloat = screen.height - bottomInset - cardHeight / 2
        let clampedY: CGFloat = maxY < minY ? minY : min(max(rawY, minY), maxY)

        VStack(alignment: .leading, spacing: 0) {
            Text(mark.title)
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(Color.navy)

            Text(mark.body)
                .font(.system(size: 12))
                .foregroundStyle(Color.navy.opacity(0.72))
                .lineSpacing(6) // 1.5 line height for 12pt
                .fixedSize(horizontal: false, vertical: true)
                .padding(.top, 4)

            HStack {
                HStack(spacing: 4) {
                    ForEach(0..<total, id: \.self) { i in
                        Circle()
                            .fill(i == index ? Color.navy : Color.navy.opacity(0.30))
                            .frame(width: 5, height: 5)
                    }
                }
                Spacer()
                Text("Skip")
                    .font(.system(size: 11))
                    .foregroundStyle(Color.navy.opacity(0.72))
                    .contentShape(Rectangle())
                    .onTapGesture { manager.skipTour() }
                Text(manager.currentIndex == manager.activeTour.count - 1 ? "Done" : "Next")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Color.navy)
                    .padding(.leading, 8)
                    .contentShape(Rectangle())
                    .onTapGesture { manager.advance() }
            }
            .padding(.top, 10)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
        .frame(width: cardWidth)
        .background(RoundedRectangle(cornerRadius: 12).fill(Color.orange))
        .fixedSize(horizontal: false, vertical: true)
        .overlay {
            GeometryReader { cardGeo in
                Color.clear.onAppear {
                    positionedCardSize = cardGeo.size
                }
                .onChange(of: cardGeo.size) { _, newSize in
                    positionedCardSize = newSize
                }
            }
        }
        .position(
            x: clampX(anchorRect: anchorRect, cardWidth: cardWidth, screen: screen),
            y: clampedY
        )
        .ignoresSafeArea()
    }

    @State private var positionedCardSize: CGSize = .init(width: 230, height: 120)

    private func clampX(anchorRect: CGRect, cardWidth: CGFloat, screen: CGRect) -> CGFloat {
        var x = anchorRect.midX - cardWidth / 2
        x = max(10 + cardWidth / 2, x)
        x = min(screen.width - 10 - cardWidth / 2, x)
        return x
    }
}

// MARK: - Pulse ring with phase animation

private struct PulseRingView: View {
    let expanded: CGRect
    let isCircular: Bool
    @State private var pulse: Bool = false

    var body: some View {
        Group {
            if isCircular {
                let dim = max(expanded.width, expanded.height)
                let r = dim / 2
                Circle()
                    .stroke(Color.orange, lineWidth: 2)
                    .frame(width: r * 2, height: r * 2)
                    .position(x: expanded.midX, y: expanded.midY)
            } else {
                RoundedRectangle(cornerRadius: 14)
                    .stroke(Color.orange, lineWidth: 2)
                    .frame(width: expanded.width, height: expanded.height)
                    .position(x: expanded.midX, y: expanded.midY)
            }
        }
        .opacity(pulse ? 0.15 : 0.9)
        .onAppear {
            withAnimation(.easeInOut(duration: 0.85).repeatForever(autoreverses: true)) {
                pulse = true
            }
        }
    }
}
